crates/wasm-qmd-parser/src/c_shim.rs

```diff
@@ -0,0 +1,237 @@
+        .unwrap()
+}
+
+/// Store a layout in the pointer, returning a pointer to where the data should be stored.
+unsafe fn store_layout(buf: *mut u8, layout: Layout, offset_to_data: usize) -> *mut c_void {
+    *(buf as *mut Layout) = layout;
+    (buf as *mut u8).offset(offset_to_data as isize) as *mut c_void
+}
+
+/* -------------------------------- string.h -------------------------------- */
+
+#[no_mangle]
+pub unsafe extern "C" fn memcpy(dest: *mut c_void, src: *const c_void, size: usize) -> *mut c_void {
+    std::ptr::copy_nonoverlapping(src, dest, size);
+    dest
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn memmove(
+    dest: *mut c_void,
+    src: *const c_void,
+    size: usize,
+) -> *mut c_void {
+    std::ptr::copy(src, dest, size);
+    dest
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn memset(s: *mut c_void, c: i32, n: usize) -> *mut c_void {
+    let slice = std::slice::from_raw_parts_mut(s as *mut u8, n);
+    slice.fill(c as u8);
+    s
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn memcmp(ptr1: *const c_void, ptr2: *const c_void, n: usize) -> c_int {
+    let s1 = std::slice::from_raw_parts(ptr1 as *const u8, n);
+    let s2 = std::slice::from_raw_parts(ptr2 as *const u8, n);
+
+    for (a, b) in s1.iter().zip(s2.iter()) {
+        if *a != *b {
+            return (*a as i32) - (*b as i32);
+        }
+    }
+
+    0
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn strncmp(ptr1: *const c_void, ptr2: *const c_void, n: usize) -> c_int {
+    let s1 = std::slice::from_raw_parts(ptr1 as *const u8, n);
+    let s2 = std::slice::from_raw_parts(ptr2 as *const u8, n);
+
+    for (a, b) in s1.iter().zip(s2.iter()) {
+        if *a != *b || *a == 0 {
+            return (*a as i32) - (*b as i32);
+        }
+    }
+
+    0
+}
+
+/* -------------------------------- wctype.h -------------------------------- */
+
+#[no_mangle]
+pub unsafe extern "C" fn iswspace(c: c_int) -> bool {
+    char::from_u32(c as u32).map_or(false, |c| c.is_whitespace())
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn iswalnum(c: c_int) -> bool {
+    char::from_u32(c as u32).map_or(false, |c| c.is_alphanumeric())
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn iswdigit(c: c_int) -> bool {
+    char::from_u32(c as u32).map_or(false, |c| c.is_digit(10))
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn iswalpha(c: c_int) -> bool {
+    char::from_u32(c as u32).map_or(false, |c| c.is_alphabetic())
+}
+
+// Note: Not provided by https://github.com/cacticouncil/lilypad, but we needed
+// this one too. We could contribute this back upstream? Note that
+// `towlower()`'s C function docs say it is only guaranteed to work in 1:1
+// mapping cases, so that is what we reimplement here as well.
+// https://en.cppreference.com/w/c/string/wide/towlower
+#[no_mangle]
+pub unsafe extern "C" fn towlower(c: c_int) -> c_int {
+    char::from_u32(c as u32).map_or(0, |c| {
+        c.to_lowercase().next().map(|c| c as i32).unwrap_or(0)
+    })
+}
```

Comment:
Nearly straight from https://github.com/cacticouncil/lilypad/blob/main/src/c_shim.rs but I did need this one, which is a blessed tree-sitter function and you use it in your scanner